### DISCUSSION TITLE
A shame I cannot get it to work

### DISCUSSION THREAD

**Discussion by realcbnewham** [estivensh4/aws-kmp#62] [General]
This looks to be a great little library and is desperately needed because who knows how long Amazon will take to publish a working version for Kotlin Native.

Unfortunately, while I followed the instructions on your website and provided the correct credentials, it doesn't work. In Android Studio it won't even build:
> Task :shared:cinteropAWSS3IosArm64
Exception in thread "main" java.lang.Error: /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS17.4.sdk/System/Library/Frameworks/Foundation.framework/Headers/NSObjCRuntime.h:615:74: error: unknown type name 'NSUInteger'

which has only happened since I added the build.gradle and POD lines in your instructions.

On Xcode the app builds but as soon as I attempt to list a bucket it throws an exception.

I look forward to seeing the library working!

**Comment by realcbnewham** [comment]
I've worked some more on this. The generatePresignedUrl() works only if I provide my original key and secret. If I use a temporary credential (which is a must-have for security reasons) then it fails. One reason is that no token is generated as part of the URL (X-Amz-Security-Token). I note that this is automatically generated on Android AWS SDK version of the generatePresignedUrl() function. If I manually add the session token from my temporary credentials to the generated URL string then it works to the extent of accepting the token but subsequently fails because the signature is not valid, the error returned being:

SignatureDoesNotMatchThe request signature we calculated does not match the signature you provided. Check your key and signing method.

The credentials given to the generate function are exactly the same as those given to my Android version and that produces the correct signature and loads the S3 contents as expected.

Lastly, I notice that adding this library causes problems with Android Studio - it's what I originally observed when I said it "didn't compile". It does - but only on every second compilation! I've tried clearing the caches, etc but that has no effect. This may be a problem with how the project is set up - I've no idea - but it's been compiling and running fine through 4 weeks of development until I added this library. The error I get is:

Exception in thread "main" java.lang.Error: /Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS17.4.sdk/System/Library/Frameworks/Foundation.framework/Headers/NSObjCRuntime.h:615:74: error: unknown type name 'NSUInteger'
/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS17.4.sdk/System/Library/Frameworks/Foundation.framework/Headers/NSObjCRuntime.h:615:104: error: unknown type name 'NSUInteger'
/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS17.4.sdk/System/Library/Frameworks/Foundation.framework/Headers/NSObjCRuntime.h:631:24: error: unknown type name 'NSInteger'
/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS17.4.sdk/System/Library/Frameworks/Foundation.framework/Headers/NSObjCRuntime.h:631:24: error: unknown type name 'NSInteger'
/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS17.4.sdk/System/Library/Frameworks/Foundation.framework/Headers/NSObjCRuntime.h:639:20: error: unknown type name 'NSUInteger'
/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS17.4.sdk/System/Library/Frameworks/Foundation.framework/Headers/NSObjCRuntime.h:639:20: error: unknown type name 'NSUInteger'
/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS17.4.sdk/System/Library/Frameworks/Foundation.framework/Headers/NSObjCRuntime.h:644:20: error: unknown type name 'NSUInteger'
/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS17.4.sdk/System/Library/Frameworks/Foundation.framework/Headers/NSObjCRuntime.h:644:20: error: unknown type name 'NSUInteger'
/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS17.4.sdk/System/Library/Frameworks/Foundation.framework/Headers/NSObjCRuntime.h:649:17: error: unknown type name 'NSInteger'
/Applications/Xcode.app/Contents/Developer/Platforms/iPhoneOS.platform/Developer/SDKs/iPhoneOS17.4.sdk/System/Library/Frameworks/Foundation.framework/Headers/NSObjCRuntime.h:649:17: error: unknown type name 'NSInteger'
	at org.jetbrains.kotlin.native.interop.indexer.ModuleSupportKt.getModulesASTFiles(ModuleSupport.kt:80)
	at org.jetbrains.kotlin.native.interop.indexer.ModuleSupportKt.getModulesInfo(ModuleSupport.kt:15)
	at org.jetbrains.kotlin.native.interop.gen.jvm.MainKt.buildNativeLibrary(main.kt:564)
	at org.jetbrains.kotlin.native.interop.gen.jvm.MainKt.processCLib(main.kt:307)
	at org.jetbrains.kotlin.native.interop.gen.jvm.MainKt.processCLibSafe(main.kt:243)
	at org.jetbrains.kotlin.native.interop.gen.jvm.MainKt.access$processCLibSafe(main.kt:1)
	at org.jetbrains.kotlin.native.interop.gen.jvm.Interop.interop(main.kt:101)
	at org.jetbrains.kotlin.cli.utilities.InteropCompilerKt.invokeInterop(InteropCompiler.kt:47)
	at org.jetbrains.kotlin.cli.utilities.MainKt.mainImpl(main.kt:23)
	at org.jetbrains.kotlin.cli.utilities.MainKt.main(main.kt:45)

**Reply by estivensh** [reply]
received @realcbnewham . I'll be checking it out, could you mount the bug type issue please?

**Reply by realcbnewham** [reply]
Thanks for the reply.  One last thing, in looking at the code I see you do:

val credentials = AWSStaticCredentialsProvider(accessKey, secretKey)

but in a Swift app I wrote 7 years ago and an Android app from a couple of years ago in Java I'm doing:

let basicCreds = AWSBasicSessionCredentialsProvider(accessKey: credentials!.accessKey, secretKey: credentials!.secretAccessKey, sessionToken: credentials!.sessionToken)

and the same general setup in Android/Java.

I'm passing the token to the configuration.  Anyway, I'm not brilliant at nosing through other people's code, but this might be the reason why the token is not being appended. No idea about the signature as that's internal to AWS, but it may require the token as part of it's setup, which could  be why it's not correct.

Anyway - keep up the good work!  I'll try and add a bug issue soon.

**Reply by estivensh** [reply]
Thanks for your contribution @realcbnewham  this will be very useful to check what is going on.
